src/dns/server.rs

```diff
@@ -89,10 +89,21 @@ impl Server {
         socket_factory: &(dyn SocketFactory + Send + Sync),
         allow_unknown_source: bool,
     ) -> Result<Self, Error> {
-        let address = address.with_ipv6(socket_factory.ipv6_enabled_localhost().unwrap_or_else(|e| {
-            warn!(err=?e, "failed to determine if IPv6 was disabled; continuing anyways, but this may fail");
-            true
-        }));
+        // if the address we got from config is supposed to be v6-enabled,
+        // actually check if the local pod context our socketfactory operates in supports V6.
+        // This is to ensure globally-enabled V6 support doesn't try to create V6 addresses in pods
+        // that do not support it, and also ensure globally-disabled V6 support doesn't create V6 address
+        // even if that's turned off.
+        let local_address = if let config::Address::Localhost(v6_global_enable, _) = address {
+            if v6_global_enable {
+                address.with_ipv6(socket_factory.ipv6_enabled_localhost().unwrap_or_else(|e| {
```

Comment:
nit: maybe we should put this in with_ipv6 and maybe call it something like "maybe disable ipv6"?

Reply:
Yeah it's a bit too cute doing it out here, agreed.